scripts/util/build-doc.js

```diff
@@ -0,0 +1,293 @@
+#!/usr/bin/env node
+
+var path = require('path');
+var fs = require('fs');
+var yuidoc = require('./yuidoc');
+var NodeGit = require('nodegit');
+var util = require('util');
+var recursiveRmdir = require('./recursive-rmdir');
+var recursiveCpdir = require('./recursive-cpdir');
+
+function buildTreePromise(commitPromise, name) {
+  return commitPromise.then(function (commit){
+    return commit.getTree();
+  }).then(function (tree){
+    return {
+      name: name,
+      tree: tree
+    }
+  });
+}
+
+/** This function checks out trees synchronously. 
+  * Important because attempting to checkout async causes errors
+  */
+function checkoutTrees(datum, rest, options, callback, paths) {
+  paths = paths || [];
+  var checkoutOptions = {
+    checkoutStrategy: options.checkoutStrategy,
+    paths: options.srcDirName,
+    targetDirectory: path.join(options.repo.workdir(), options.apiDirPath, datum.name)
+  };
+
+  NodeGit.Checkout.tree(options.repo, datum.tree, checkoutOptions).then(function () {
+    paths.push(checkoutOptions.targetDirectory)
+    if (rest.length) {
+      checkoutTrees(rest.shift(), rest, options, callback, paths);
+    } else if (callback) {
+      callback(paths);
+    }
+  }).catch(function (error){
+    error.message = `Unable to checkout "${datum.name}" | ${error.message}`
+    console.error(error);
+  });
+}
+
+function DocBuilder(options) {
+  this.options = util._extend({
+    masterBranchName: 'master',
+    docsBranchName: 'gh-pages',
+    docsDirName: 'docs',
+    versionsDirName: 'versions',
+    srcDirName: 'src',
+    rmSrcDir: false,
+    commitDocs: false,
+    checkoutStrategy: ( NodeGit.Checkout.STRATEGY.FORCE + NodeGit.Checkout.STRATEGY.DONT_WRITE_INDEX ),
+  }, options);
+
+  if (!this.options.themeDir) {
+    this.options.themeDir = path.join(this.options.docsDirName, 'yuidoc', 'theme');
+  }
+
+  this.options.apiDirPath = path.join(this.options.docsDirName, this.options.versionsDirName);
+}
+
+DocBuilder.prototype.checkoutDocsBranch = function (callback) {
+  var self = this;
+  if (fs.existsSync(self.options.docsDirName)) {
+    recursiveRmdir(self.options.docsDirName);
+  }
+
+  fs.mkdirSync(self.options.docsDirName)
+
+  var repo;
+
+  console.info(`\nChecking out ${self.options.docsBranchName} branch`);
+  NodeGit.Repository.open('.').then(function (_repo) {
+    repo = _repo;
```

Comment:
This is an ugly way to capture the param. You should push this through with what's resolved, or properly nest your promise scopes.